File: service/game/actions/conflict.py
from service.game.actions.base import Command


class ContestDevelopmentCommand(Command):
    def execute(self, game_state, player):
        if game_state.phase != "WORK":
            return False
        if player.health in ["sick", "recovering"]:
            return False

        dev_id = self.payload.get("dev_id")
        side = self.payload.get("side")
        development = game_state.developments.get(dev_id)
        if not development:
            return False

        if side == "INITIATOR":
            if (
                development.is_contested
                or development.pending_contest
                or development.owner == player.session_id
            ):
                return False
            development.pending_contest = True
            development.contest_initiator_id = player.session_id
            development.pending_contest_day = game_state.day + 1
            player.add_timeline_event(
                "ACTION_COMPLETED",
                {"action": "CONTEST_SCHEDULED", "dev_id": dev_id},
            )
            game_state.contest_count += 1
            return True

        if side == "CONTESTER":
            if not development.is_contested:
                return False
            if player.session_id not in development.contester_supporters:
                development.contester_supporters.append(player.session_id)
        elif side == "OWNER":
            if not development.is_contested:
                return False
            if player.session_id not in development.owner_supporters:
                development.owner_supporters.append(player.session_id)

        player.committed_action = {
            "type": "CONTEST_ACTION",
            "dev_id": dev_id,
            "side": side,
        }
        player.add_timeline_event(
            "ACTION_COMPLETED",
            {"action": "CONTEST", "dev_id": dev_id, "side": side},
        )
        return True


def activate_pending_contests(game_state):
    for development in game_state.developments.values():
        if (
            getattr(development, "pending_contest", False)
            and development.pending_contest_day == game_state.day
        ):
            development.is_contested = True
            development.contester_supporters = []
            development.pending_contest = False
            owner = game_state.players.get(development.owner)
            if owner:
                owner.add_timeline_event(
                    "CONTEST_STARTED",
                    {
                        "dev_id": development.id,
                        "attacker": development.contest_initiator_id,
                    },
                )


def resolve_contests(game_state):
    for development in game_state.developments.values():
        if not getattr(development, "is_contested", False):
            continue

        development.contester_supporters = []
        development.owner_supporters = []
        for player in game_state.players.values():
            action = getattr(player, "committed_action", None)
            if (
                action
                and isinstance(action, dict)
                and action.get("type") == "CONTEST_ACTION"
                and action.get("dev_id") == development.id
            ):
                side = action.get("side")
                if side == "CONTESTER":
                    development.contester_supporters.append(
                        player.session_id
                    )
                elif side == "OWNER":
                    development.owner_supporters.append(player.session_id)

        contester_score = len(development.contester_supporters)
        owner_score = len(development.owner_supporters)
        contester_present = (
            development.contest_initiator_id
            in development.contester_supporters
        )
        owner_present = development.owner in development.owner_supporters

        if not contester_present:
            development.is_contested = False
            development.contest_initiator_id = None
        elif not owner_present or contester_score > owner_score:
            old_owner = game_state.players.get(development.owner)
            new_owner = game_state.players.get(
                development.contest_initiator_id
            )
            if old_owner and development.id in old_owner.developments:
                old_owner.developments.remove(development.id)
            if new_owner and development.id not in new_owner.developments:
                new_owner.developments.append(development.id)
            development.owner = development.contest_initiator_id
            development.is_contested = False
            development.contest_initiator_id = None
        elif owner_score > contester_score:
            development.is_contested = False
            development.contest_initiator_id = None

        development.contester_supporters = []
        development.owner_supporters = []

File: service/src/game/actions/conflict.ts
import {
  contestDevPayloadSchema,
  type CommittedContestActionDTO,
  type DevelopmentDTO,
  type PlayerDTO,
} from "@takes-a-village/shared";

export interface ContestDevelopment extends DevelopmentDTO {
  pendingContestDay?: number;
}

interface ContestPlayer {
  readonly id: string;
  readonly developments: string[];
  readonly timeline: PlayerDTO["timeline"];
  health: PlayerDTO["health"];
  committedAction: PlayerDTO["committed_action"];
  finishedPhase: boolean;
}

interface ContestGame {
  readonly players: Map<string, ContestPlayer>;
  readonly developments: Map<string, ContestDevelopment>;
  readonly day: number;
  readonly phase: "WORK" | "TRADE" | "NIGHT";
}

export function handleContestAction(
  game: ContestGame,
  player: ContestPlayer,
  command: unknown,
  payload: Record<string, unknown>,
): boolean {
  if (command !== "CONTEST_DEV" || game.phase !== "WORK" || player.health === "sick" || player.health === "recovering") {
    return false;
  }
  const parsed = contestDevPayloadSchema.safeParse(payload);
  if (!parsed.success) return false;
  const development = game.developments.get(parsed.data.dev_id);
  if (!development) return false;

  if (parsed.data.side !== "INITIATOR") {
    if (!development.is_contested || (parsed.data.side !== "CONTESTER" && parsed.data.side !== "OWNER")) return false;
    const supporters = parsed.data.side === "CONTESTER"
      ? development.contester_supporters
      : development.owner_supporters;
    if (!supporters?.includes(player.id)) supporters?.push(player.id);
    player.committedAction = {
      type: "CONTEST_ACTION",
      dev_id: development.id,
      side: parsed.data.side,
    };
    player.timeline.push({
      type: "ACTION_COMPLETED",
      data: { action: "CONTEST", dev_id: development.id, side: parsed.data.side },
    });
    player.finishedPhase = true;
    return true;
  }

  if (development.is_contested
    || development.pending_contest
    || development.owner_id === player.id) return false;

  development.pending_contest = true;
  development.pendingContestDay = game.day + 1;
  development.contest_initiator_id = player.id;
  player.timeline.push({
    type: "ACTION_COMPLETED",
    data: { action: "CONTEST_SCHEDULED", dev_id: development.id },
  });
  player.finishedPhase = true;
  return true;
}

export function activatePendingContests(game: ContestGame): void {
  for (const development of game.developments.values()) {
    if (!development.pending_contest || development.pendingContestDay !== game.day) continue;
    development.is_contested = true;
    development.contester_supporters = [];
    development.pending_contest = false;
    const owner = game.players.get(development.owner_id);
    if (owner) {
      owner.timeline.push({
        type: "CONTEST_STARTED",
        data: { dev_id: development.id, attacker: development.contest_initiator_id ?? null },
      });
    }
  }
}

export function resolveContests(game: ContestGame): void {
  for (const development of game.developments.values()) {
    if (!development.is_contested) continue;
    development.contester_supporters = [];
    development.owner_supporters = [];
    for (const player of game.players.values()) {
      const action = asContestAction(player.committedAction);
      if (!action || action.dev_id !== development.id) continue;
      if (action.side === "CONTESTER") development.contester_supporters.push(player.id);
      else development.owner_supporters.push(player.id);
    }

    const attackerId = development.contest_initiator_id;
    const attackerPresent = typeof attackerId === "string" && development.contester_supporters.includes(attackerId);
    const ownerPresent = development.owner_supporters.includes(development.owner_id);
    if (!attackerPresent) endContest(development);
    else if (!ownerPresent || development.contester_supporters.length > development.owner_supporters.length) {
      const oldOwner = game.players.get(development.owner_id);
      const newOwner = game.players.get(attackerId);
      if (oldOwner) removeValue(oldOwner.developments, development.id);
      if (newOwner && !newOwner.developments.includes(development.id)) newOwner.developments.push(development.id);
      development.owner_id = attackerId;
      endContest(development);
    }
    else if (development.owner_supporters.length > development.contester_supporters.length) endContest(development);

    development.contester_supporters = [];
    development.owner_supporters = [];
  }
}

function asContestAction(action: PlayerDTO["committed_action"]): CommittedContestActionDTO | null {
  if (!action || !("type" in action) || action.type !== "CONTEST_ACTION") return null;
  return action;
}

function endContest(development: ContestDevelopment): void {
  development.is_contested = false;
  development.contest_initiator_id = null;
}

function removeValue(values: string[], value: string): void {
  const index = values.indexOf(value);
  if (index >= 0) values.splice(index, 1);
}
